### PR TITLE
Implement second-level card tables for non-Windows platforms

### PR DESCRIPTION
**Pull request by adityamandaleeka**
This change enables us to have card-bundles on platforms where we don't have the ability to use OS WriteWatch APIs. For now, it's implemented for AMD64 only.

The first commit does some misc clean up including fixing broken tabular formatting caused by a global find and replace a while back, adding a tiny amount of documentation/comments for things that are non-obvious, and renaming some variables like x/y/z to more meaningful names. This is by no means a full clean-up, but it does improve the readability of the card table/bundle related code.

I will run a few extra CI jobs on this, so please do not merge yet.

### CODE REVIEW

File: clrdefinitions.cmake [hunk 1]
```diff
@@ -189,6 +189,7 @@ if(CLR_CMAKE_PLATFORM_UNIX_AMD64)
 endif (CLR_CMAKE_PLATFORM_UNIX_AMD64)
 add_definitions(-DFEATURE_USE_ASM_GC_WRITE_BARRIERS)
 if(CLR_CMAKE_PLATFORM_ARCH_AMD64 AND NOT WIN32)
+  # add_definitions(-DFEATURE_MANUALLY_MANAGED_CARD_BUNDLES)
```
Comment (adityamandaleeka) [inline clrdefinitions.cmake]: I will un-comment this. I'm making sure that everything is good with and without this feature defined.

**Comment by adityamandaleeka** [comment]
@Maoni0 @swgillespie

**Comment by Maoni0** [comment]
I'd reviewed the changes with Aditya (reviewed GC changes in detail, not so much about the write barrier changes). looping in @PeterSolMS to see if he has any objections to the write barrier changes.

**Comment by Maoni0** [comment]
hmm, I guess Peter is already on vacation :P @jkotas could you please review the write barrier changes in this PR?

File: src/vm/amd64/jithelpers_fast.S [hunk 2]
```diff
@@ -303,16 +339,36 @@ LEAF_ENTRY JIT_ByRefWriteBarrier, _TEXT
...
     UpdateCardTable_ByRefWriteBarrier:
-        mov     byte ptr [rcx], 0FFh
+        mov     byte ptr [rcx + rax], 0FFh
+
+#ifdef FEATURE_MANUALLY_MANAGED_CARD_BUNDLES
+        // Shift rcx by 0x0A more to get the card bundle byte (we shifted by 0x0B already)
+        shr     rcx, 0x0A
+        
+        PREPARE_EXTERNAL_VAR g_card_bundle_table, rax
+        mov     rax, [rax]
```
Comment (adityamandaleeka) [inline src/vm/amd64/jithelpers_fast.S]: Yup, good idea 👍

**Comment by adityamandaleeka** [comment]
@dotnet-bot test OSX Release longgc

File: src/vm/amd64/jithelpers_fast.S [hunk 3]
```diff
@@ -163,12 +179,32 @@ LEAF_ENTRY JIT_WriteBarrier, _TEXT
...
-        // jne     UpdateCardTable
         .byte 0x75, 0x02
+        // jne     UpdateCardTable
         REPRET
 
     UpdateCardTable:
         mov     byte ptr [rdi + rax], 0FFh
+
+#ifdef FEATURE_MANUALLY_MANAGED_CARD_BUNDLES
+        NOP_3_BYTE // padding for alignment of constant
```
Comment (jkotas) [inline src/vm/amd64/jithelpers_fast.S]: There is 6 byte nop instruction that would better here - check the Intel manual:

```
66 NOP DWORD ptr [EAX + EAX*1 + 00H]
Byte sequence: 66 0F 1F 44 00 00H
```

File: src/vm/amd64/jithelpers_fastwritebarriers.S [hunk 4]
```diff
@@ -40,6 +45,27 @@ PATCH_LABEL JIT_WriteBarrier_PreGrow64_Patch_Label_CardTable
 
     UpdateCardTable_PreGrow64:
         mov     byte ptr [rdi + rax], 0FFh
+
+#ifdef FEATURE_MANUALLY_MANAGED_CARD_BUNDLES
+        NOP_3_BYTE // padding for alignment of constant
```
Comment (jkotas) [inline src/vm/amd64/jithelpers_fastwritebarriers.S]: Same here

File: src/vm/amd64/jithelpers_fast.S (hunk 3 above)
Comment (adityamandaleeka) [inline src/vm/amd64/jithelpers_fast.S]: Sounds good, thanks.

**Comment by adityamandaleeka** [comment]
@swgillespie @jkotas I've implemented the changes you suggested. PTAL.

Please don't merge this yet... I'll merge it myself.

**Comment by swgillespie** [comment]
argh - whatever I just did, consider it a LGTM 👍

**Comment by adityamandaleeka** [comment]
@jkotas Would you like more time to review this PR?

**Pull request merged**